In [1]:
import tensorflow as tf

In [2]:
tf.__version__

'2.15.0'

In [3]:
tf.random.set_seed(42)

model_cifar = tf.keras.Sequential()

model_cifar.add(tf.keras.layers.Flatten(input_shape=(32, 32, 3)))

for _ in range(20):
    model_cifar.add(
        tf.keras.layers.Dense(100, activation="swish", kernel_initializer="he_normal")
    )

2024-05-18 19:59:17.879616: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M1 Pro
2024-05-18 19:59:17.879661: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 16.00 GB
2024-05-18 19:59:17.879680: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 5.33 GB
2024-05-18 19:59:17.879748: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:306] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2024-05-18 19:59:17.879794: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:272] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)


In [11]:
model_cifar.add(tf.keras.layers.Dense(10, activation="softmax"))

In [13]:
model_cifar.compile(loss="sparse_categorical_crossentropy", optimizer=tf.keras.optimizers.legacy.Nadam(learning_rate=5e-5), metrics=["accuracy"])

In [4]:
cifar10 = tf.keras.datasets.cifar10.load_data()

In [5]:
(X_train_full, y_train_full), (X_test, y_test) = cifar10

X_train = X_train_full[5000:]
y_train = y_train_full[5000:]
X_valid = X_train_full[:5000]
y_valid = y_train_full[:5000]

In [6]:
from pathlib import Path

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
model_ckpt_cb = tf.keras.callbacks.ModelCheckpoint("model_cifar10.ckpt", save_best_only=True)

run_idx = 1
run_logdir = Path() / "cifar10_logs" / f"run_{run_idx:03d}"
tensorboard_cb = tf.keras.callbacks.TensorBoard(run_logdir)

In [7]:
callback_lst = [early_stopping_cb, model_ckpt_cb, tensorboard_cb]

In [9]:
%load_ext tensorboard
%tensorboard --logdir=./cifar10_logs

In [14]:
model_cifar.fit(X_train, y_train, epochs=100, validation_data=(X_valid, y_valid), callbacks=callback_lst, verbose=1)

Epoch 1/100


2024-05-18 20:15:12.068478: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


1407/1407 [==============================] - ETA: 0s - loss: 3.7540 - accuracy: 0.1688INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 67s 42ms/step - loss: 3.7540 - accuracy: 0.1688 - val_loss: 2.4276 - val_accuracy: 0.1818
Epoch 2/100
1406/1407 [============================>.] - ETA: 0s - loss: 2.0881 - accuracy: 0.2368INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 2.0881 - accuracy: 0.2368 - val_loss: 2.1398 - val_accuracy: 0.2162
Epoch 3/100
1407/1407 [==============================] - ETA: 0s - loss: 1.9619 - accuracy: 0.2832INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.9619 - accuracy: 0.2832 - val_loss: 2.0066 - val_accuracy: 0.2522
Epoch 4/100
1407/1407 [==============================] - ETA: 0s - loss: 1.8804 - accuracy: 0.3158INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 57s 41ms/step - loss: 1.8804 - accuracy: 0.3158 - val_loss: 1.9375 - val_accuracy: 0.3178
Epoch 5/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.8199 - accuracy: 0.3410INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.8199 - accuracy: 0.3410 - val_loss: 1.8636 - val_accuracy: 0.3350
Epoch 6/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.7634 - accuracy: 0.3623INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 57s 40ms/step - loss: 1.7633 - accuracy: 0.3623 - val_loss: 1.7778 - val_accuracy: 0.3616
Epoch 7/100
1407/1407 [==============================] - ETA: 0s - loss: 1.7145 - accuracy: 0.3820INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.7145 - accuracy: 0.3820 - val_loss: 1.7666 - val_accuracy: 0.3464
Epoch 8/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.6768 - accuracy: 0.3983INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 59s 42ms/step - loss: 1.6768 - accuracy: 0.3983 - val_loss: 1.6723 - val_accuracy: 0.4070
Epoch 9/100
1407/1407 [==============================] - 56s 39ms/step - loss: 1.6419 - accuracy: 0.4086 - val_loss: 1.6758 - val_accuracy: 0.3976
Epoch 10/100
1407/1407 [==============================] - 56s 40ms/step - loss: 1.6150 - accuracy: 0.4184 - val_loss: 1.6799 - val_accuracy: 0.3908
Epoch 11/100
1407/1407 [==============================] - 58s 41ms/step - loss: 1.5918 - accuracy: 0.4284 - val_loss: 1.6827 - val_accuracy: 0.3940
Epoch 12/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5718 - accuracy: 0.4378INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 59s 42ms/step - loss: 1.5718 - accuracy: 0.4378 - val_loss: 1.6036 - val_accuracy: 0.4254
Epoch 13/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.5512 - accuracy: 0.4422INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 60s 43ms/step - loss: 1.5512 - accuracy: 0.4422 - val_loss: 1.6017 - val_accuracy: 0.4230
Epoch 14/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5331 - accuracy: 0.4521INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 58s 41ms/step - loss: 1.5331 - accuracy: 0.4521 - val_loss: 1.6009 - val_accuracy: 0.4272
Epoch 15/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.5167 - accuracy: 0.4568INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 58s 41ms/step - loss: 1.5168 - accuracy: 0.4568 - val_loss: 1.5651 - val_accuracy: 0.4386
Epoch 16/100
1407/1407 [==============================] - 57s 41ms/step - loss: 1.4987 - accuracy: 0.4648 - val_loss: 1.5809 - val_accuracy: 0.4344
Epoch 17/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.4886 - accuracy: 0.4667INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 57s 40ms/step - loss: 1.4886 - accuracy: 0.4668 - val_loss: 1.5548 - val_accuracy: 0.4474
Epoch 18/100
1407/1407 [==============================] - 54s 38ms/step - loss: 1.4733 - accuracy: 0.4730 - val_loss: 1.5647 - val_accuracy: 0.4430
Epoch 19/100
1406/1407 [============================>.] - ETA: 0s - loss: 1.4627 - accuracy: 0.4757INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.4627 - accuracy: 0.4758 - val_loss: 1.5461 - val_accuracy: 0.4464
Epoch 20/100
1407/1407 [==============================] - 54s 38ms/step - loss: 1.4503 - accuracy: 0.4806 - val_loss: 1.5738 - val_accuracy: 0.4372
Epoch 21/100
1407/1407 [==============================] - ETA: 0s - loss: 1.4381 - accuracy: 0.4836INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.4381 - accuracy: 0.4836 - val_loss: 1.5460 - val_accuracy: 0.4490
Epoch 22/100
1407/1407 [==============================] - ETA: 0s - loss: 1.4281 - accuracy: 0.4892INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.4281 - accuracy: 0.4892 - val_loss: 1.5151 - val_accuracy: 0.4650
Epoch 23/100
1407/1407 [==============================] - 54s 38ms/step - loss: 1.4163 - accuracy: 0.4946 - val_loss: 1.5434 - val_accuracy: 0.4394
Epoch 24/100
1407/1407 [==============================] - 54s 39ms/step - loss: 1.4055 - accuracy: 0.4978 - val_loss: 1.5603 - val_accuracy: 0.4410
Epoch 25/100
1407/1407 [==============================] - ETA: 0s - loss: 1.3944 - accuracy: 0.5030INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10.ckpt/assets


1407/1407 [==============================] - 56s 40ms/step - loss: 1.3944 - accuracy: 0.5030 - val_loss: 1.4975 - val_accuracy: 0.4666
Epoch 26/100
1407/1407 [==============================] - 54s 38ms/step - loss: 1.3845 - accuracy: 0.5054 - val_loss: 1.5520 - val_accuracy: 0.4494
Epoch 27/100
1407/1407 [==============================] - 55s 39ms/step - loss: 1.3747 - accuracy: 0.5103 - val_loss: 1.5065 - val_accuracy: 0.4678
Epoch 28/100
1407/1407 [==============================] - 70s 50ms/step - loss: 1.3694 - accuracy: 0.5130 - val_loss: 1.5098 - val_accuracy: 0.4634
Epoch 29/100
1407/1407 [==============================] - 124s 88ms/step - loss: 1.3596 - accuracy: 0.5145 - val_loss: 1.5136 - val_accuracy: 0.4636
Epoch 30/100
1407/1407 [==============================] - 55s 39ms/step - loss: 1.3477 - accuracy: 0.5180 - val_loss: 1.6237 - val_accuracy: 0.4454
Epoch 31/100
1407/1407 [==============================] - 55s 39ms/step - loss: 1.3427 - accuracy: 0.5211 - val_loss: 1.5900

In [15]:
model_cifar.evaluate(X_valid, y_valid)

157/157 [==============================] - 1s 7ms/step - loss: 1.4975 - accuracy: 0.4666


[1.4974884986877441, 0.4666000008583069]

### Model with BatchNormalization

In [17]:
model_cifar_2 = tf.keras.Sequential()

model_cifar_2.add(tf.keras.layers.Flatten(input_shape=(32, 32, 3)))

for _ in range(20):
    model_cifar_2.add(tf.keras.layers.Dense(100, kernel_initializer="he_normal"))
    model_cifar_2.add(tf.keras.layers.BatchNormalization())
    model_cifar_2.add(tf.keras.layers.Activation("swish"))

model_cifar_2.add(tf.keras.layers.Dense(10, activation="softmax"))

In [18]:
model_cifar_2.compile(loss="sparse_categorical_crossentropy", optimizer=tf.keras.optimizers.legacy.Nadam(learning_rate=5e-4), metrics=["accuracy"])

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
model_ckpt_cb = tf.keras.callbacks.ModelCheckpoint("model_cifar10_BN.ckpt", save_best_only=True)

run_idx = 1
run_logdir = Path() / "cifar10_BN_logs" / f"run_{run_idx:03d}"
tensorboard_cb = tf.keras.callbacks.TensorBoard(run_logdir)

In [19]:
model_cifar_2.fit(X_train, y_train, epochs=100,
          validation_data=(X_valid, y_valid),
          callbacks=[early_stopping_cb, model_ckpt_cb, tensorboard_cb])

Epoch 1/100
1407/1407 [==============================] - ETA: 0s - loss: 2.0516 - accuracy: 0.2493INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 154s 94ms/step - loss: 2.0516 - accuracy: 0.2493 - val_loss: 1.9194 - val_accuracy: 0.3210
Epoch 2/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.7941 - accuracy: 0.3541 - val_loss: 1.9652 - val_accuracy: 0.2990
Epoch 3/100
1407/1407 [==============================] - ETA: 0s - loss: 1.6885 - accuracy: 0.3957INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 126s 90ms/step - loss: 1.6885 - accuracy: 0.3957 - val_loss: 1.7603 - val_accuracy: 0.3510
Epoch 4/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.6176 - accuracy: 0.4205 - val_loss: 1.7798 - val_accuracy: 0.3562
Epoch 5/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5627 - accuracy: 0.4438INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 201s 143ms/step - loss: 1.5627 - accuracy: 0.4438 - val_loss: 1.6296 - val_accuracy: 0.4134
Epoch 6/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5125 - accuracy: 0.4626INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 537s 382ms/step - loss: 1.5125 - accuracy: 0.4626 - val_loss: 1.5547 - val_accuracy: 0.4526
Epoch 7/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.4639 - accuracy: 0.4801 - val_loss: 1.6329 - val_accuracy: 0.4204
Epoch 8/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.4256 - accuracy: 0.4946 - val_loss: 1.5887 - val_accuracy: 0.4356
Epoch 9/100
1407/1407 [==============================] - ETA: 0s - loss: 1.3925 - accuracy: 0.5089INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 126s 89ms/step - loss: 1.3925 - accuracy: 0.5089 - val_loss: 1.5417 - val_accuracy: 0.4560
Epoch 10/100
1407/1407 [==============================] - ETA: 0s - loss: 1.3602 - accuracy: 0.5188INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 126s 90ms/step - loss: 1.3602 - accuracy: 0.5188 - val_loss: 1.5251 - val_accuracy: 0.4602
Epoch 11/100
1407/1407 [==============================] - ETA: 0s - loss: 1.3341 - accuracy: 0.5265INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 128s 91ms/step - loss: 1.3341 - accuracy: 0.5265 - val_loss: 1.5029 - val_accuracy: 0.4746
Epoch 12/100
1407/1407 [==============================] - 124s 88ms/step - loss: 1.3014 - accuracy: 0.5418 - val_loss: 1.7394 - val_accuracy: 0.4026
Epoch 13/100
1407/1407 [==============================] - 121s 86ms/step - loss: 1.2806 - accuracy: 0.5477 - val_loss: 1.5940 - val_accuracy: 0.4458
Epoch 14/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.2523 - accuracy: 0.5583 - val_loss: 1.5170 - val_accuracy: 0.4642
Epoch 15/100
1407/1407 [==============================] - ETA: 0s - loss: 1.2286 - accuracy: 0.5641INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_BN.ckpt/assets


1407/1407 [==============================] - 126s 90ms/step - loss: 1.2286 - accuracy: 0.5641 - val_loss: 1.4798 - val_accuracy: 0.4938
Epoch 16/100
1407/1407 [==============================] - 124s 88ms/step - loss: 1.2170 - accuracy: 0.5664 - val_loss: 1.4881 - val_accuracy: 0.4818
Epoch 17/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.1898 - accuracy: 0.5807 - val_loss: 1.5437 - val_accuracy: 0.4674
Epoch 18/100
1407/1407 [==============================] - 122s 87ms/step - loss: 1.1703 - accuracy: 0.5906 - val_loss: 1.6278 - val_accuracy: 0.4376
Epoch 19/100
1407/1407 [==============================] - 124s 88ms/step - loss: 1.1566 - accuracy: 0.5943 - val_loss: 1.6980 - val_accuracy: 0.4228
Epoch 20/100
1407/1407 [==============================] - 124s 88ms/step - loss: 1.1379 - accuracy: 0.6014 - val_loss: 1.5466 - val_accuracy: 0.4652
Epoch 21/100
1407/1407 [==============================] - 121s 86ms/step - loss: 1.1241 - accuracy: 0.6054 - val_loss: 

In [20]:
model_cifar_2.evaluate(X_valid, y_valid)

157/157 [==============================] - 2s 10ms/step - loss: 1.4798 - accuracy: 0.4938


[1.479790210723877, 0.49380001425743103]

### Model with Dropout and it's variants

In [21]:
model_cifar_3 = tf.keras.Sequential()
model_cifar_3.add(tf.keras.layers.Flatten(input_shape=(32, 32, 3)))

for _ in range(20):
    model_cifar_3.add(tf.keras.layers.Dense(100, kernel_initializer="he_normal", activation="swish"))

model_cifar_3.add(tf.keras.layers.Dropout(rate=0.2))
model_cifar_3.add(tf.keras.layers.Dense(10, activation="softmax"))

In [22]:
model_cifar_3.compile(loss="sparse_categorical_crossentropy", optimizer=tf.keras.optimizers.legacy.Nadam(learning_rate=5e-4), metrics=["accuracy"])

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
model_ckpt_cb = tf.keras.callbacks.ModelCheckpoint("model_cifar10_Dropout.ckpt", save_best_only=True)

run_idx = 1
run_logdir = Path() / "cifar10_Dropout_logs" / f"run_{run_idx:03d}"
tensorboard_cb = tf.keras.callbacks.TensorBoard(run_logdir)

In [23]:
model_cifar_3.fit(X_train, y_train, epochs=100,
          validation_data=(X_valid, y_valid),
          callbacks=[early_stopping_cb, model_ckpt_cb, tensorboard_cb])

Epoch 1/100
1407/1407 [==============================] - ETA: 0s - loss: 3.4996 - accuracy: 0.1857INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 68s 43ms/step - loss: 3.4996 - accuracy: 0.1857 - val_loss: 2.0639 - val_accuracy: 0.2206
Epoch 2/100
1407/1407 [==============================] - ETA: 0s - loss: 1.9770 - accuracy: 0.2596INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 58s 41ms/step - loss: 1.9770 - accuracy: 0.2596 - val_loss: 1.9412 - val_accuracy: 0.2714
Epoch 3/100
1407/1407 [==============================] - 55s 39ms/step - loss: 1.9104 - accuracy: 0.2894 - val_loss: 1.9439 - val_accuracy: 0.2736
Epoch 4/100
1407/1407 [==============================] - 56s 40ms/step - loss: 1.8645 - accuracy: 0.3094 - val_loss: 2.0662 - val_accuracy: 0.2592
Epoch 5/100
1407/1407 [==============================] - ETA: 0s - loss: 1.8238 - accuracy: 0.3317INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 58s 41ms/step - loss: 1.8238 - accuracy: 0.3317 - val_loss: 1.8384 - val_accuracy: 0.3302
Epoch 6/100
1407/1407 [==============================] - 56s 40ms/step - loss: 1.7842 - accuracy: 0.3486 - val_loss: 2.0090 - val_accuracy: 0.2940
Epoch 7/100
1407/1407 [==============================] - ETA: 0s - loss: 1.7442 - accuracy: 0.3653INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 58s 41ms/step - loss: 1.7442 - accuracy: 0.3653 - val_loss: 1.7443 - val_accuracy: 0.3736
Epoch 8/100
1407/1407 [==============================] - ETA: 0s - loss: 1.7163 - accuracy: 0.3794INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 54s 38ms/step - loss: 1.7163 - accuracy: 0.3794 - val_loss: 1.6807 - val_accuracy: 0.3910
Epoch 9/100
1407/1407 [==============================] - 44s 31ms/step - loss: 1.7319 - accuracy: 0.3787 - val_loss: 1.7275 - val_accuracy: 0.3830
Epoch 10/100
1407/1407 [==============================] - 44s 32ms/step - loss: 1.6861 - accuracy: 0.3922 - val_loss: 1.6924 - val_accuracy: 0.3844
Epoch 11/100
1407/1407 [==============================] - 44s 31ms/step - loss: 1.6585 - accuracy: 0.3993 - val_loss: 1.7676 - val_accuracy: 0.3624
Epoch 12/100
1407/1407 [==============================] - ETA: 0s - loss: 1.6434 - accuracy: 0.4075INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 46s 32ms/step - loss: 1.6434 - accuracy: 0.4075 - val_loss: 1.6382 - val_accuracy: 0.4200
Epoch 13/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.6182 - accuracy: 0.4188 - val_loss: 1.6707 - val_accuracy: 0.3960
Epoch 14/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.6029 - accuracy: 0.4239 - val_loss: 1.6544 - val_accuracy: 0.4184
Epoch 15/100
1407/1407 [==============================] - 47s 33ms/step - loss: 1.5883 - accuracy: 0.4300 - val_loss: 1.6590 - val_accuracy: 0.3940
Epoch 16/100
1407/1407 [==============================] - 44s 31ms/step - loss: 1.7332 - accuracy: 0.3702 - val_loss: 1.6658 - val_accuracy: 0.3878
Epoch 17/100
1407/1407 [==============================] - 46s 32ms/step - loss: 1.6414 - accuracy: 0.4095 - val_loss: 1.6733 - val_accuracy: 0.3926
Epoch 18/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.6162 - accuracy: 0.4169 - val_loss: 1.6580 

INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 46s 33ms/step - loss: 1.5879 - accuracy: 0.4279 - val_loss: 1.6268 - val_accuracy: 0.4208
Epoch 20/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.5697 - accuracy: 0.4343 - val_loss: 1.6385 - val_accuracy: 0.4080
Epoch 21/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5537 - accuracy: 0.4421INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 47s 33ms/step - loss: 1.5537 - accuracy: 0.4421 - val_loss: 1.5892 - val_accuracy: 0.4284
Epoch 22/100
1407/1407 [==============================] - ETA: 0s - loss: 1.5346 - accuracy: 0.4493INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 47s 33ms/step - loss: 1.5346 - accuracy: 0.4493 - val_loss: 1.5759 - val_accuracy: 0.4348
Epoch 23/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.5217 - accuracy: 0.4547 - val_loss: 1.6120 - val_accuracy: 0.4234
Epoch 24/100
1407/1407 [==============================] - 46s 32ms/step - loss: 1.5115 - accuracy: 0.4589 - val_loss: 1.5825 - val_accuracy: 0.4348
Epoch 25/100
1407/1407 [==============================] - ETA: 0s - loss: 1.4987 - accuracy: 0.4639INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 47s 34ms/step - loss: 1.4987 - accuracy: 0.4639 - val_loss: 1.5584 - val_accuracy: 0.4496
Epoch 26/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4861 - accuracy: 0.4684 - val_loss: 1.5808 - val_accuracy: 0.4322
Epoch 27/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4770 - accuracy: 0.4726 - val_loss: 1.5805 - val_accuracy: 0.4368
Epoch 28/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4665 - accuracy: 0.4752 - val_loss: 1.5901 - val_accuracy: 0.4278
Epoch 29/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4644 - accuracy: 0.4755 - val_loss: 1.5606 - val_accuracy: 0.4488
Epoch 30/100
1407/1407 [==============================] - 47s 33ms/step - loss: 1.5043 - accuracy: 0.4625 - val_loss: 1.6383 - val_accuracy: 0.4086
Epoch 31/100
1407/1407 [==============================] - ETA: 0s - loss: 1.4636 - accuracy: 0.4766INFO:tensorflow:Assets wri

INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 49s 35ms/step - loss: 1.4636 - accuracy: 0.4766 - val_loss: 1.5374 - val_accuracy: 0.4564
Epoch 32/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4358 - accuracy: 0.4857 - val_loss: 1.5639 - val_accuracy: 0.4462
Epoch 33/100
1407/1407 [==============================] - 47s 33ms/step - loss: 1.4254 - accuracy: 0.4877 - val_loss: 1.5416 - val_accuracy: 0.4490
Epoch 34/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4211 - accuracy: 0.4898 - val_loss: 1.5868 - val_accuracy: 0.4458
Epoch 35/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.4123 - accuracy: 0.4930 - val_loss: 1.5512 - val_accuracy: 0.4542
Epoch 36/100
1407/1407 [==============================] - 47s 34ms/step - loss: 1.4075 - accuracy: 0.4953 - val_loss: 1.5617 - val_accuracy: 0.4546
Epoch 37/100
1407/1407 [==============================] - ETA: 0s - loss: 1.3989 - accuracy: 0.4989INFO:tensorflow:Assets wri

INFO:tensorflow:Assets written to: model_cifar10_Dropout.ckpt/assets


1407/1407 [==============================] - 47s 34ms/step - loss: 1.3989 - accuracy: 0.4989 - val_loss: 1.5208 - val_accuracy: 0.4616
Epoch 38/100
1407/1407 [==============================] - 46s 33ms/step - loss: 1.3924 - accuracy: 0.5021 - val_loss: 1.5561 - val_accuracy: 0.4520
Epoch 39/100
1407/1407 [==============================] - 46s 32ms/step - loss: 1.3821 - accuracy: 0.5037 - val_loss: 1.5479 - val_accuracy: 0.4480
Epoch 40/100
1407/1407 [==============================] - 46s 32ms/step - loss: 1.3773 - accuracy: 0.5054 - val_loss: 1.5615 - val_accuracy: 0.4522
Epoch 41/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.3696 - accuracy: 0.5096 - val_loss: 1.5719 - val_accuracy: 0.4550
Epoch 42/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.3673 - accuracy: 0.5123 - val_loss: 1.5377 - val_accuracy: 0.4606
Epoch 43/100
1407/1407 [==============================] - 45s 32ms/step - loss: 1.3582 - accuracy: 0.5122 - val_loss: 1.5585 

In [31]:
model_cifar_3.evaluate(X_valid, y_valid)

157/157 [==============================] - 1s 6ms/step - loss: 2.3027 - accuracy: 0.0972


[2.3027231693267822, 0.09719999879598618]

### Summary
- Model 1: `model_cifar`; `validation acc`: `46.66%`
- Model 2: `model_cifar_2`: `validation acc`: `49.38%`
- Model 3: `model_cifar_3`: `validation acc`: `9.72%`


In [24]:
class MCDropout(tf.keras.layers.Dropout):
    def call(self, inputs):
        return super().call(inputs, training=True)
    
model_mcd = tf.keras.Sequential([
    MCDropout(layer.rate) if isinstance(layer, tf.keras.layers.Dropout)
    else layer for layer in model_cifar_3.layers
])

In [25]:
def mc_dropout_predict_probas(mc_model, X, n_samples=1000):
    Y_probas = [mc_model.predict(X) for sample in range(n_samples)]
    return np.mean(Y_probas, axis=0)

def mc_dropout_predict_classes(mc_model, X, n_samples=1000):
    Y_probas = mc_dropout_predict_probas(mc_model, X, n_samples)
    return Y_probas.argmax(axis=1)

In [ ]:
tf.random.set_seed(42)

y_pred = mc_dropout_predict_classes(model_mcd, X_valid)
accuracy = (y_pred == y_valid[:, 0]).mean()
accuracy

### Utilities

The `find_learning_rate()` function trains the model using the `ExponentialLearningRate` callback, and it returns the learning rates and corresponding batch losses. 
At the end, it restores the model and its optimizer to their initial state.

In [26]:
import math
K = tf.keras.backend

class ExponentialLearningRate(tf.keras.callbacks.Callback):
    def __init__(self, factor):
        self.factor = factor
        self.rates = []
        self.losses = []

    def on_epoch_begin(self, epoch, logs=None):
        self.sum_of_epoch_losses = 0

    def on_batch_end(self, batch, logs=None):
        mean_epoch_loss = logs["loss"]  # the epoch's mean loss so far 
        new_sum_of_epoch_losses = mean_epoch_loss * (batch + 1)
        batch_loss = new_sum_of_epoch_losses - self.sum_of_epoch_losses
        self.sum_of_epoch_losses = new_sum_of_epoch_losses
        self.rates.append(K.get_value(self.model.optimizer.learning_rate))
        self.losses.append(batch_loss)
        K.set_value(self.model.optimizer.learning_rate,
                    self.model.optimizer.learning_rate * self.factor)

def find_learning_rate(model, X, y, epochs=1, batch_size=32, min_rate=1e-4,
                       max_rate=1):
    init_weights = model.get_weights()
    iterations = math.ceil(len(X) / batch_size) * epochs
    factor = (max_rate / min_rate) ** (1 / iterations)
    init_lr = K.get_value(model.optimizer.learning_rate)
    K.set_value(model.optimizer.learning_rate, min_rate)
    exp_lr = ExponentialLearningRate(factor)
    history = model.fit(X, y, epochs=epochs, batch_size=batch_size,
                        callbacks=[exp_lr])
    K.set_value(model.optimizer.learning_rate, init_lr)
    model.set_weights(init_weights)
    return exp_lr.rates, exp_lr.losses

The `OneCycleScheduler` custom callback updates the learning rate at the beginning of each batch. It applies the logic: increase the learning rate linearly during about half of training, then reduce it linearly back to the initial learning rate, and lastly reduce it down to close to zero linearly for the very last part of training.

In [27]:
class OneCycleScheduler(tf.keras.callbacks.Callback):
    def __init__(self, iterations, max_lr=1e-3, start_lr=None,
                 last_iterations=None, last_lr=None):
        self.iterations = iterations
        self.max_lr = max_lr
        self.start_lr = start_lr or max_lr / 10
        self.last_iterations = last_iterations or iterations // 10 + 1
        self.half_iteration = (iterations - self.last_iterations) // 2
        self.last_lr = last_lr or self.start_lr / 1000
        self.iteration = 0

    def _interpolate(self, iter1, iter2, lr1, lr2):
        return (lr2 - lr1) * (self.iteration - iter1) / (iter2 - iter1) + lr1

    def on_batch_begin(self, batch, logs):
        if self.iteration < self.half_iteration:
            lr = self._interpolate(0, self.half_iteration, self.start_lr,
                                   self.max_lr)
        elif self.iteration < 2 * self.half_iteration:
            lr = self._interpolate(self.half_iteration, 2 * self.half_iteration,
                                   self.max_lr, self.start_lr)
        else:
            lr = self._interpolate(2 * self.half_iteration, self.iterations,
                                   self.start_lr, self.last_lr)
        self.iteration += 1
        K.set_value(self.model.optimizer.learning_rate, lr)

In [28]:
n_epochs = 50
batch_size = 128

n_iterations = math.ceil(len(X_train) / batch_size) * n_epochs
onecycle = OneCycleScheduler(n_iterations, max_lr=0.05)
history = model_cifar_3.fit(X_train, y_train, epochs=n_epochs, batch_size=batch_size,
                    validation_data=(X_valid, y_valid),
                    callbacks=[onecycle])

Epoch 1/50
352/352 [==============================] - 16s 37ms/step - loss: 22.8247 - accuracy: 0.1200 - val_loss: 2.3029 - val_accuracy: 0.1010
Epoch 2/50
352/352 [==============================] - 12s 33ms/step - loss: 2.3029 - accuracy: 0.1004 - val_loss: 2.3032 - val_accuracy: 0.0972
Epoch 3/50
352/352 [==============================] - 11s 32ms/step - loss: 2.3031 - accuracy: 0.0996 - val_loss: 2.3031 - val_accuracy: 0.1010
Epoch 4/50
352/352 [==============================] - 11s 32ms/step - loss: 2.3032 - accuracy: 0.0968 - val_loss: 2.3027 - val_accuracy: 0.0996
Epoch 5/50
352/352 [==============================] - 11s 32ms/step - loss: 2.3033 - accuracy: 0.0999 - val_loss: 2.3032 - val_accuracy: 0.1010
Epoch 6/50
352/352 [==============================] - 11s 32ms/step - loss: 2.3037 - accuracy: 0.0980 - val_loss: 2.3030 - val_accuracy: 0.0920
Epoch 7/50
352/352 [==============================] - 11s 32ms/step - loss: 2.3035 - accuracy: 0.0976 - val_loss: 2.3057 - val_accuracy

In [30]:
# history2 = model_cifar_2.fit(X_train, y_train, epochs=n_epochs, batch_size=batch_size,
#                     validation_data=(X_valid, y_valid),
#                     callbacks=[onecycle])